### ISSUE TITLE
Implement excision

### ISSUE THREAD

**Issue by ncalexan** [mozilla/datomish#21]
See http://docs.datomic.com/excision.html.

**Comment by rnewman** [comment]
Requirements/description:

Excision cannot remove vocabulary, the bootstrap schema, or any other built-ins. This includes the `:db/excise` attributes themselves.

Excision records datoms that define the excision itself, and additionally removes the affected datoms from history. Indeed, recording these datoms is how one triggers an excision, though I'm sure we'll wrap this in a nice API.

Excision targets _either_ an entity or an attribute. (If you want to treat an attribute as an entity, you need to specify attributes to excise.)

If an entity is targeted, all _inbound_ and _outbound_ datoms are removed, as are any component entities linked by an outbound datom.

An optional set of attributes can be specified to filter the set of affected datoms.

An optional timestamp or transaction ID filter can be specified, similarly.

If an attribute is targeted, it affect all entities, subject to filtering as above.

It's important that excision is deterministic: for a given database state, a given excision input should always remove the exact same set of datoms.

**Comment by rnewman** [comment]
Implementation choices:

This is likely to hook into [the same spot in the transactor that picks up schema changes](https://github.com/mozilla/mentat/blob/master/db/src/tx.rs#L715):

```rust
        if tx_might_update_metadata {
            // Extract changes to metadata from the store.
            let metadata_assertions = self.store.committed_metadata_assertions(self.tx_id)?;
```

I suggest that at that point we synchronously compute an 'excision operation' that can be persisted in the database (ensuring that all excisions eventually and atomically take place), in a separate table, and then process that table after the transact commits. This is inline with Datomic's background behavior. We can expose a synchronous 'purge' call that blocks until all pending excisions are processed.

Excision will mutate the datoms table, the log, potentially the fulltext values table (Datomic doesn't support excision of fulltext, but I think we should), and update the cache. It'll have no interaction with syncing _except_ that certain sync operations — particularly merges — should only proceed against a store with no pending excisions, for obvious determinism reasons.

**Comment by ncalexan** [comment]
> It'll have no interaction with syncing except that certain sync operations — particularly merges — should only proceed against a store with no pending excisions, for obvious determinism reasons.

I started thinking about this a little after our conversation yesterday, @rnewman, and it seems like excision should interact with syncing.  Suppose that excision leaves data on the server: then a user in possession of their credentials could be compelled to provide them, and the data on the server could be recovered.  That's a different guarantee than we provide when deleting data right now, and it's a weaker guarantee.

Server-side excision would require some assistance from the client, and would have a very different work-load.  It might look like a bulk replacement of transactions in-place, rather than the append-only workload we're currently expecting.

Thoughts?

**Comment by rnewman** [comment]
> … it seems like excision should interact with syncing.

Yes, in ways that are already documented elsewhere — in the above I meant that local asynchronous excision markers don't really interact with local sync code.

Excision and remote timelines are naturally related: any rewritten transaction chunks need to be uploaded, their old transaction entries rewritten to point to the new chunks, and the old chunks flagged for deletion or garbage collection.

This is append-only for data chunks (modulo flagging for cleanup), and read-write for the transaction metadata entries themselves.

See "Rewriting history" in [my protocol outline](https://public.etherpad-mozilla.org/p/esp).
